google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/Util.java

```diff
@@ -148,4 +151,44 @@ static Long getGfeLatency(Metadata metadata) {
     }
     return null;
   }
+
+  private static boolean getLocationFromMetadata(
```

Comment:
I think you should clean this up to avoid side effects:

```
ResponseParams getResponseParams(@Nullable Metadata m) {
    if (m == null) {
        return null;
    }
    try {
        return ResponseParams.parseForm(...)
    }
    return null;
}

recordMetricsFromMetadata((...) {
    // ...
    // Depending on the environment ResponseParams can either be in headers or trailers
    // check headers first
    ResponseParams rp = getResponseParams(responseMetadata.getMetadata());
    // fall back on trailers
    if (rp == null) {
        rp = getResponseParams(responseMetadata.getTrailingMetadata())
    }
    if (rp != null) {
        tracer.setLocations(decodedTrailers.getZoneId(), decodedTrailers.getClusterId());
    } else if (rp == null && latency == null) {
        latency = 0
    }

}
```

Reply:
or you can do it via optionals:
Optional<ResponseParams> getResponseParams(@Nullable Metadata m) {
    if (m == null) {
        return Optional.empty();
    }
    try {
        return Optional.of(ResponseParams.parseForm(...))
    }
    return null;
}

recordMetricsFromMetadata((...) {
    // ...
    // Depending on the environment ResponseParams can either be in headers or trailers
    // check headers first
    Optional<ResponseParams> rp = 
        getResponseParams(responseMetadata.getMetadata());
        .orElse(() -> getResponseParams(responseMetadata.getTrailingMetadata()));
    
    rp.ifPresent((rp) => tracer.setLocations(rp.getZoneId(), rp.getClusterId());)
    
    if (rp.isEmpty() && latency == null) {
        latency = 0
    }

}